### ISSUE TITLE
[ONNX] Dynamic shape is not respected for Tensor.expand

### ISSUE THREAD

**Issue by icyblade** [pytorch/pytorch#87677]
#### Description
Dynamic shape is not respected for op `Tensor.expand`.
For repo below, both axis 0 (batch_size) and 1 (sequence_length) are dynamic axis. However, when processing `position_ids.expand(input_ids.size())`, dynamic shapes are fixed based on input sample. When inferencing, shape error will be raised.

#### Workaround
`Tensor.repeat` instead of `Tensor.expand`.

#### Repro
```python
import onnxruntime
import torch
import torch.nn as nn
import torch.onnx


class TestModel(nn.Module):
    def forward(self, input_ids):
        position_ids = torch.arange(input_ids.size(1), dtype=torch.long).unsqueeze(0)
        position_ids = position_ids.expand(input_ids.size())  # Issue here
        # position_ids = position_ids.repeat(input_ids.size(0), 1)  # Workaround
        return position_ids
    
torch.onnx.export(
    TestModel(),
    torch.ones(1, 512),
    'tmp.onnx',
    input_names=['input_ids'],
    output_names=['output'],
    dynamic_axes={
        'input_ids': {0: 'batch_size', 1: 'sequence_length'},
        'output': {0: 'batch_size'}
    },
    opset_version=14,
)

sess = onnxruntime.InferenceSession('tmp.onnx')
print(torch.__version__)
print(onnxruntime.__version__)
print(sess.run(None, {'input_ids': [[0, 1]]})[0].shape)
print(sess.run(None, {'input_ids': [[0, 1]] * 1000})[0].shape)
```

#### Error message
```
2022-10-25 05:47:13.308286660 [W:onnxruntime:, execution_frame.cc:812 VerifyOutputSizes] Expected shape from model of {batch_size,512} does not match actual shape of {1,2} for output output
2022-10-25 05:47:13.308821123 [W:onnxruntime:, execution_frame.cc:812 VerifyOutputSizes] Expected shape from model of {batch_size,512} does not match actual shape of {1000,2} for output output
```

### Versions

```
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-194-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 515.76
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.971
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] pytorch-lightning==1.6.5
[pip3] torch==1.11.0
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py38h7f8727e_0  
[conda] mkl_fft                   1.3.1            py38hd3c417c_0  
[conda] mkl_random                1.2.2            py38h51133e4_0  
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] pytorch                   1.11.0          py3.8_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-lightning         1.6.5                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchmetrics              0.9.3                    pypi_0    pypi
[conda] torchvision               0.12.0               py38_cu113    pytorch
```

**Comment by justinchuby** [comment]
Thanks, It does look like expand is not respecting the input as a tensor when it should:

https://github.com/pytorch/pytorch/blob/41d8502c9aeca4b115f3cca8c26e43cee6dee9c5/torch/onnx/symbolic_opset9.py#L876-L891

cc @BowenBao @AllenTiTaiWang

**Comment by icyblade** [comment]
Thanks @justinchuby for the confirmation. 

In the meanwhile, we also observe a ~2x performance degrade for the wordaround, probably because `expand` creates a view while `repeat` copies the tensor.

### Benchmark code
```python
import torch
import torch.nn as nn
import torch.utils.benchmark as benchmark

class Model0(nn.Module):
    def forward(self, input_ids):
        position_ids = torch.arange(input_ids.size(1), dtype=torch.long).unsqueeze(0)
        position_ids = position_ids.expand(input_ids.size())
        # position_ids = position_ids.repeat(input_ids.size(0), 1)
        return position_ids

class Model1(nn.Module):
    def forward(self, input_ids):
        position_ids = torch.arange(input_ids.size(1), dtype=torch.long).unsqueeze(0)
        # position_ids = position_ids.expand(input_ids.size())
        position_ids = position_ids.repeat(input_ids.size(0), 1)
        return position_ids
    
sample_input = torch.ones(32, 512, dtype=torch.long)

t0 = benchmark.Timer(
    stmt='model(sample_input)',
    setup='model=Model().eval().cuda()',
    globals={'sample_input': sample_input, 'Model': Model0}
)
t1 = benchmark.Timer(
    stmt='model(sample_input)',
    setup='model=Model().eval().cuda()',
    globals={'sample_input': sample_input, 'Model': Model1}
)
print(t0.timeit(100000))
print(t1.timeit(100000))
```

### Result
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7f8e241b68e0>
model(sample_input)
setup: model=Model().eval().cuda()
  7.21 us
  1 measurement, 100000 runs , 1 thread
<torch.utils.benchmark.utils.common.Measurement object at 0x7f8dfd3a5fd0>
model(sample_input)
setup: model=Model().eval().cuda()
  13.10 us
  1 measurement, 100000 runs , 1 thread
```

**Comment by AllenTiTaiWang** [comment]
Is the repro supposed to show error messages in onnxruntime step? I don't get error with nightly. Could you try nightly?

**Comment by icyblade** [comment]
Thanks @AllenTiTaiWang. Confirmed this bug is fixed in `1.14.0.dev20221114+cu116`. To get the benefit, should user upgrade to 1.14.0 (after released)? Will the fix get backported to versions <1.14.0?

**Comment by AllenTiTaiWang** [comment]
ONNX related bugs fixed are always merged into master branch only. There won't be a fix < 1.14.0.
